crates/wasi/src/p3/cli/host.rs

```diff
@@ -1,155 +1,119 @@
     environment, exit, stderr, stdin, stdout, terminal_input, terminal_output, terminal_stderr,
     terminal_stdin, terminal_stdout,
 };
 use crate::p3::cli::{TerminalInput, TerminalOutput};
-use crate::p3::{DEFAULT_BUFFER_CAPACITY, MAX_BUFFER_CAPACITY};
 use anyhow::{Context as _, anyhow};
 use bytes::BytesMut;
+use core::pin::Pin;
+use core::task::{Context, Poll};
 use std::io::Cursor;
-use std::pin::Pin;
-use std::task::{self, Context, Poll};
-use tokio::io::{AsyncRead, AsyncReadExt as _, AsyncWrite, AsyncWriteExt as _, ReadBuf};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
 use wasmtime::StoreContextMut;
 use wasmtime::component::{
     Accessor, Destination, Resource, Source, StreamConsumer, StreamProducer, StreamReader,
     StreamResult,
 };
 
-struct InputStreamProducer<T> {
-    rx: T,
+struct InputStreamProducer {
+    rx: Pin<Box<dyn AsyncRead + Send + Sync>>,
 }
 
-impl<T, D> StreamProducer<D> for InputStreamProducer<T>
-where
-    T: AsyncRead + Send + Unpin + 'static,
-{
+impl<D> StreamProducer<D> for InputStreamProducer {
     type Item = u8;
     type Buffer = Cursor<BytesMut>;
 
     fn poll_produce<'a>(
-        self: Pin<&mut Self>,
+        mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
         store: StoreContextMut<'a, D>,
-        destination: &'a mut Destination<'a, Self::Item, Self::Buffer>,
+        dst: &'a mut Destination<'a, Self::Item, Self::Buffer>,
         finish: bool,
     ) -> Poll<wasmtime::Result<StreamResult>> {
-        if finish {
-            return Poll::Ready(Ok(StreamResult::Cancelled));
-        }
-
-        let me = self.get_mut();
-
-        Poll::Ready(Ok(
-            if let Some(mut destination) = destination.as_direct_destination(store)
-                && !destination.remaining().is_empty()
-            {
-                let mut buffer = ReadBuf::new(destination.remaining());
-                match task::ready!(Pin::new(&mut me.rx).poll_read(cx, &mut buffer)) {
-                    Ok(()) => {
-                        if buffer.filled().is_empty() {
-                            StreamResult::Dropped
-                        } else {
-                            let count = buffer.filled().len();
-                            destination.mark_written(count);
-                            StreamResult::Completed
-                        }
-                    }
-                    Err(_) => {
-                        // TODO: Report the error to the guest
-                        StreamResult::Dropped
-                    }
+        if let Some(mut dst) = dst.as_direct_destination(store) {
+            let mut buf = ReadBuf::new(dst.remaining());
+            match self.rx.as_mut().poll_read(cx, &mut buf) {
+                Poll::Ready(Ok(())) if buf.capacity() == 0 => {
+                    Poll::Ready(Ok(StreamResult::Completed))
                 }
-            } else {
-                let capacity = destination
-                    .remaining(store)
```

Comment:
this would always return `None`, wouldn't it? (since that's a guest buffer)

Reply:
Good point; I didn't revisit that code after I added the `if/else` block around it.